### DISCUSSION TITLE
Large format machine - some thoughts about enhancements

### DISCUSSION THREAD

**Discussion by phil-barrett** [terjeio/ioSender#120] [Ideas]
I have been working on getting the Avid Pro4848 (1250mm x 1250mm) running with grblHAL and ioSender.  Over all, it has been a positive experience though not without hiccups.  From this effort, I have a few ideas to make ioSender better for big CNC routers.

1. Allow configuration of the jog tab's speed and distance.  (Not keyboard jogging which is configurable)  On my Pro4848, I can do rapids of up to 25000 mm/min.  With the 10 mm and 1000 mm/min max, it takes forever to get to the other side. I wind up typing G0 commands instead. I would prefer to not have to touch the keyboard.
2. In the 3D view tab, the spindle/bit is so small it is impossible to see. Please make the size configurable. (Yes, I know others have asked, too)  Maybe make the default size proportional to the program limits?
3. In the 3D tab, moire' becomes a bigger issue and is highly distracting.  Not sure what can easily be done about it but would be really nice.
4. In the Console tab, add a "smart verbose" filter that displays only changed idle messages. Even better, allow specifying which fields to watch for change.  In setting up limit sensors, it would be nice to display only idle messages where the limit switches changed. Since the machine is fairly big, I have to move around to actually test trigger them and the idle messages have scrolled off the screen. I have to uncheck verbose and scroll back.

**Comment by terjeio** [comment]
1. Is the number of radio buttons for selection ok? I need 16 new config variables for the current 8 (mm and inches mode are different).
2. Better to scale according to zoom?
3. You have to get a higher resolution monitor?
4. Filter out unchanged responses (the whole string) is easy. Adding a custom filter would have to be done by using regular expressions or grok as mentioned in #66? Too complicated to add in a meaningful way?

**Reply by phil-barrett** [reply]
1. I don't think more than 4 selections are needed. 
2. Scale to zoom would be the best.
3. My monitor is set up to 1920x1024.  Bought it last week for $150 USD. Pretty standard resolution. I know you use a package for the graphics display, maybe it's just not possible to fix (easily, anyway). 
4. Yeah, I didn't expect filtering for specific fields to be easy.  I would be happy with just unchanged.

By the way, your link to that console enhancement thread reminded me of something.  Having a dedicated computer with a large display (Intel NUC, 1920x1024 monitor), I find myself maximizing (i.e. full screen) ioSender.  As such having a persistent jog panel would be really nice.

**Reply by terjeio** [reply]
1 and 4 are now available for test in the latest [edge version](http://www.io-engineering.com/downloads). Note that I have not yet added imperial mode for 1.

3. I do not believe it is possible to fix this short of analyzing the gcode and remove enough paths from the display to make it disappear.

**Comment by phil-barrett** [comment]
wow, fast.

I don't think solving moire' problem takes GCode reduction.  It looks like it is due to the background color?
![image](https://user-images.githubusercontent.com/16570958/127062830-2e07e174-11ea-4a1a-a928-9d58823f4339.png)

**Reply by phil-barrett** [reply]
Ran edge4,  the two changes look good.

**Reply by terjeio** [reply]
>  It looks like it is due to the background color?

It is the grid? Switch it off to check. The grid should be scaled with zoom as well...

**Reply by phil-barrett** [reply]
yup, it's the grid.  is there a way to specify the grid spacing - seems way too fine.

**Reply by terjeio** [reply]
> is there a way to specify the grid spacing

No. I want to adjust it dynamically on zoom level. I am looking for an event to subscribe to but have not yet found it.

**Comment by unit16v** [comment]
I'm building a large format polishing machine using IoSender and GrblHal on a Teensy fitted to one of Phil's boards and issue 2 has been a problem for me too.
The tool position indicator on the monitor is too small for my size of workpiece (1250 x 600mm).
![tool position indicator](https://user-images.githubusercontent.com/90523384/137633980-e720e4f0-ed5e-4277-a14f-7172dd48a5b8.png)
I'm not a programmer, so don't bite if I've done it wrong, but I've managed to make the indicator bigger by changing line 311 in  the Visual Studio Solution 'CNC GCodeViewer - Renderer.Xaml.cs' from: 
tool.TopRadius = tool.Height / 5d;
to:
tool.TopRadius = tool.Height * 1d;
![tool position indicator revised](https://user-images.githubusercontent.com/90523384/137634347-5077acf2-9376-4849-9a23-470fc32c4c87.png)

In an ideal world, it would be nice to have a variable assigned to the indicator and maybe a box in the settings so the user could vary the indicator size depending on the job on the table. In the meantime, you might find my solution helpful.

**Reply by terjeio** [reply]
> In an ideal world, it would be nice to have a variable assigned to the indicator and maybe a box in the settings so the user could vary the indicator size depending on the job on the table. In the meantime, you might find my solution helpful.

I have made an attempt to resize it automatically based on zoom level (and also change the grid) but I could not get it to work, it could be that using the job size (program limits) would be easier - I'll look into that.

**Reply by unit16v** [reply]
Using the job size (program limits)  to automatically resize the indicator sounds like the best way. 
In the meantime, for my installation, I’ve added a box to the Gcode Viewer panel ("Tool indicator scale") so I can input a scaling factor for the tool indicator size. 
Below are images of two jobs of different sizes and the lines of code I added. 
In my workshop, I need to see the screen from a few metres away and the red dot is very important!

![75 per cent](https://user-images.githubusercontent.com/90523384/139426326-821c35f5-1f2e-4516-9a7e-67f37ddce26a.png)
![500 per cent](https://user-images.githubusercontent.com/90523384/139426354-58796542-7ccc-45a5-8d7c-a6fc4073f28b.png)

\CNC Controls\AppConfig.cs(110): private int _toolIndScale = 100;
\CNC Controls\AppConfig.cs(117): public int ToolIndScale { get { return _toolIndScale; } set { _toolIndScale = value; OnPropertyChanged(); } }
\CNC GCodeViewer\ConfigControl.xaml(15): <Controls:NumericField x:Uid="fld_toolIndScale" Label="Tool indicator scale:" Unit="%" Format="###" ColonAt="130" Value="{Binding Path=GCodeViewer.ToolIndScale, Mode=TwoWay}" IsEnabled="{Binding Path=GCodeViewer.IsEnabled, Mode=OneWay}" />
\RenderControl.xaml.cs(72): gcodeView.ToolIndScale = AppConfig.Settings.GCodeViewer.ToolIndScale;
\CNC GCodeViewer\Renderer.xaml.cs(269): ToolIndScale = AppConfig.Settings.GCodeViewer.ToolIndScale;
\CNC GCodeViewer\Renderer.xaml.cs(292): public int ToolIndScale { get; set; } = 100;
\CNC GCodeViewer\Renderer.xaml.cs(313): tool.TopRadius = tool.Height * ToolIndScale / 200;
\CNC GCodeViewer\Renderer.xaml.cs(498): tool.TopRadius = tool.Height * ToolIndScale / 200;
\CNC GCodeViewer\Viewer.xaml.cs(96): gcodeView.ToolIndScale = AppConfig.Settings.GCodeViewer.ToolIndScale;

By the way. As a beginner to CNC I’ve found your program easy to use and it’s enabled me to get my project off the drawing board.
Thank you.
Steve R
